File: cc_sessions/hooks/session-start.py
#!/usr/bin/env python3

# ===== IMPORTS ===== #

## ===== STDLIB ===== ##
from importlib.metadata import version, PackageNotFoundError
import requests, json, sys, shutil
##-##

## ===== 3RD-PARTY ===== ##
##-##

## ===== LOCAL ===== ##
from shared_state import edit_state, PROJECT_ROOT, load_config
##-##

#-#

# ===== GLOBALS ===== #
sessions_dir = PROJECT_ROOT / 'sessions'

STATE = None
CONFIG = load_config()

developer_name = CONFIG.environment.developer_name

# Initialize context
context = f"You are beginning a new context window with the developer, {developer_name}.\n\n"

# Quick configuration checks
needs_setup = False
quick_checks = []
#-#

"""
╔═══════════════════════════════════════════════════════════════════════════════════════════════╗
║ ██████╗██████╗██████╗██████╗██████╗ █████╗ ██╗  ██╗      ██████╗██████╗ █████╗ █████╗ ██████╗ ║
║ ██╔═══╝██╔═══╝██╔═══╝██╔═══╝╚═██╔═╝██╔══██╗███╗ ██║      ██╔═══╝╚═██╔═╝██╔══██╗██╔═██╗╚═██╔═╝ ║
║ ██████╗█████╗ ██████╗██████╗  ██║  ██║  ██║████╗██║      ██████╗  ██║  ███████║█████╔╝  ██║   ║
║ ╚═══██║██╔══╝ ╚═══██║╚═══██║  ██║  ██║  ██║██╔████║      ╚═══██║  ██║  ██╔══██║██╔═██╗  ██║   ║
║ ██████║██████╗██████║██████║██████╗╚█████╔╝██║╚███║      ██████║  ██║  ██║  ██║██║ ██║  ██║   ║
║ ╚═════╝╚═════╝╚═════╝╚═════╝╚═════╝ ╚════╝ ╚═╝ ╚══╝      ╚═════╝  ╚═╝  ╚═╝  ╚═╝╚═╝ ╚═╝  ╚═╝   ║
╚═══════════════════════════════════════════════════════════════════════════════════════════════╝
SessionStart Hook

Initializes session state and loads task context:
- Checks for required components (daic command, tiktoken)  
- Clears session warning flags and stale state
- Loads current task or lists available tasks
- Updates task status from pending to in-progress
"""

# ===== EXECUTION ===== #

#!> 1. Clear flags and todos for new session
with edit_state() as s: 
    s.flags.clear_flags()
    s.todos.clear_active()
    restored = s.todos.restore_stashed()
    STATE = s
context += "Cleared session flags and active todos for new session.\n\n"

if restored:
    context += f"""Restored {restored} stashed todos from previous session:\n\n{STATE.todos.active}\n\nTo clear, use `cd .claude/hooks && python -c \"from shared_state import edit_state; with edit_state() as s: s.todos.clear_stashed()\"`\n\n"""
#!<

#!> 2. Nuke transcripts dir
transcripts_dir = sessions_dir / 'transcripts'
if transcripts_dir.exists(): shutil.rmtree(transcripts_dir, ignore_errors=True)
#!<

#!> 3. Load current task or list available tasks
# Check for active task
if (task_file := STATE.current_task.file_path) and task_file.exists():
    # Check if task status is pending and update to in-progress
    task_content = task_file.read_text()
    task_updated = False

    # Parse task frontmatter to check status
    if task_content.startswith('---'):
        lines = task_content.split('\n')
        for i, line in enumerate(lines[1:], 1):
            if line.startswith('---'):
                break
            if line.startswith('status: pending'):
                lines[i] = 'status: in-progress'
                task_updated = True
                # Write back the updated content
                task_file.write_text('\n'.join(lines))
                task_content = '\n'.join(lines)
                break

        # Output the full task state
        context += f"""Current task state:
```json
{json.dumps(STATE.current_task.task_state, indent=2)}
```

Loading task file: {STATE.current_task.file}
{"=" * 60}
{task_content}
{"=" * 60}
"""

        if task_updated: context += "[Note: Task status updated from 'pending' to 'in-progress']\n\nFollow the task-startup protocol to create branches and set up the work environment.\n\n"
        else: context += "Review the Work Log at the end of the task file above and continue the task.\n\n"
else:
    # No active task - list available tasks
    tasks_dir = sessions_dir / 'tasks'
    task_files = []

    if tasks_dir.exists(): task_files = sorted([f for f in tasks_dir.glob('*.md') if f.name != 'TEMPLATE.md'])
    for task_dir in sorted([d for d in tasks_dir.iterdir() if d.is_dir() and d.name != 'done']):
        readme_file = task_dir / 'README.md'
        if readme_file.exists(): task_files.append(task_dir)
        subtask_files = sorted([f for f in task_dir.glob('*.md') if f.name not in ['TEMPLATE.md', 'README.md']])
        task_files.extend(subtask_files)

    if task_files:
        context += "No active task set. Available tasks:\n"
        for task_file in task_files:
            fpath = task_file / 'README.md' if task_file.is_dir() else task_file
            if not fpath.exists(): continue
            # Read first few lines to get task info
            with fpath.open('r', encoding='utf-8') as f: lines = f.readlines()[:10]
            task_name = f"{task_file.name}/" if task_file.is_dir() else task_file.name
            status = 'unknown'
            for line in lines:
                if line.startswith('status:'): status = line.split(':')[1].strip(); break
            context += f"  • {task_name} ({status})\n"

        context += """
To select a task:
1. Update sessions/state/current-task.json with the task name
2. Or create a new task following sessions/protocols/task-creation.md
"""
    else: context += """No tasks found. 

To create your first task:
1. Copy the template: cp sessions/tasks/TEMPLATE.md sessions/tasks/[priority]-[task-name].md
Priority prefixes: h- (high), m- (medium), l- (low), ?- (investigate)
2. Fill in the task details
3. Update sessions/state/current-task.json
4. Follow sessions/protocols/task-startup.md
"""
#!<

#!> 4. Check cc-sessions version
try: current_version = version('cc-sessions')
except PackageNotFoundError: current_version = None
try:
    resp = requests.get("https://pypi.org/pypi/cc-sessions/json", timeout=2)
    if resp.ok:
        latest_version = resp.json().get("info", {}).get("version")
        if current_version and current_version != latest_version:
            context += f"Update available for cc-sessions: {current_version} → {latest_version}. Run `pip install --upgrade cc-sessions` to update."
except requests.RequestException: pass
#!<

#-#

output = {
    "hookSpecificOutput": {
        "hookEventName": "SessionStart",
        "additionalContext": context
    }
}
print(json.dumps(output))

sys.exit(0)

File: cc_sessions/hooks/session_start.js
#!/usr/bin/env node

// ===== IMPORTS ===== //

/// ===== STDLIB ===== ///
const fs = require('fs');
const path = require('path');
const https = require('https');
///-///

/// ===== 3RD-PARTY ===== ///
///-///

/// ===== LOCAL ===== ///
const { editState, PROJECT_ROOT, loadConfig, SessionsProtocol, getTaskFilePath, isDirectoryTask } = require('./shared_state.js');
///-///

//-//

// ===== GLOBALS ===== //
const sessionsDir = path.join(PROJECT_ROOT, 'sessions');
let STATE = null;
const CONFIG = loadConfig();
const developerName = CONFIG.environment?.developer_name || 'developer';

// Initialize context
let context = `You are beginning a new context window with the developer, ${developerName}.\n\n`;

// Quick configuration checks
let needsSetup = false;
const quickChecks = [];
//-//

/*
╔═══════════════════════════════════════════════════════════════════════════════════════════════╗
║ ██████╗██████╗██████╗██████╗██████╗ █████╗ ██╗  ██╗      ██████╗██████╗ █████╗ █████╗ ██████╗ ║
║ ██╔═══╝██╔═══╝██╔═══╝██╔═══╝╚═██╔═╝██╔══██╗███╗ ██║      ██╔═══╝╚═██╔═╝██╔══██╗██╔═██╗╚═██╔═╝ ║
║ ██████╗█████╗ ██████╗██████╗  ██║  ██║  ██║████╗██║      ██████╗  ██║  ███████║█████╔╝  ██║   ║
║ ╚═══██║██╔══╝ ╚═══██║╚═══██║  ██║  ██║  ██║██╔████║      ╚═══██║  ██║  ██╔══██║██╔═██╗  ██║   ║
║ ██████║██████╗██████║██████║██████╗╚█████╔╝██║╚███║      ██████║  ██║  ██║  ██║██║ ██║  ██║   ║
║ ╚═════╝╚═════╝╚═════╝╚═════╝╚═════╝ ╚════╝ ╚═╝ ╚══╝      ╚═════╝  ╚═╝  ╚═╝  ╚═╝╚═╝ ╚═╝  ╚═╝   ║
╚═══════════════════════════════════════════════════════════════════════════════════════════════╝
SessionStart Hook

Initializes session state and loads task context:
- Checks for required components (daic command, tiktoken)
- Clears session warning flags and stale state
- Loads current task or lists available tasks
- Updates task status from pending to in-progress
*/

// ===== FUNCTIONS ===== //

async function readStdin() {
    return new Promise((resolve) => {
        let data = '';
        process.stdin.setEncoding('utf8');
        process.stdin.on('data', chunk => data += chunk);
        process.stdin.on('end', () => {
            try {
                resolve(JSON.parse(data));
            } catch (e) {
                resolve({});
            }
        });
    });
}

function parseIndexFile(indexPath) {
    if (!fs.existsSync(indexPath)) {
        return null;
    }

    try {
        const content = fs.readFileSync(indexPath, 'utf8');
        const lines = content.split('\n');

        // Extract frontmatter using simple string parsing (like task files)
        const metadata = {};
        if (lines.length > 0 && lines[0] === '---') {
            for (let i = 1; i < lines.length; i++) {
                if (lines[i] === '---') {
                    break;
                }
                if (lines[i].includes(':')) {
                    const [key, ...valueParts] = lines[i].split(':');
                    metadata[key.trim()] = valueParts.join(':').trim();
                }
            }
        }

        // Extract task lines (those starting with - `)
        const taskLines = [];
        for (const line of lines) {
            if (line.trim().startsWith('- `')) {
                taskLines.push(line.trim());
            }
        }

        return [metadata, taskLines];
    } catch (error) {
        return null;
    }
}

function listOpenTasksGrouped() {
    const tasksDir = path.join(PROJECT_ROOT, 'sessions', 'tasks');
    const indexesDir = path.join(tasksDir, 'indexes');

    // First collect all tasks as before
    const taskFiles = [];
    if (fs.existsSync(tasksDir)) {
        const files = fs.readdirSync(tasksDir)
            .filter(f => f.endsWith('.md') && f !== 'TEMPLATE.md')
            .sort();
        taskFiles.push(...files);
    }

    // Check subdirectories
    if (fs.existsSync(tasksDir)) {
        const dirs = fs.readdirSync(tasksDir)
            .filter(item => {
                const itemPath = path.join(tasksDir, item);
                return fs.statSync(itemPath).isDirectory() && !['done', 'indexes'].includes(item);
            })
            .sort();

        for (const dir of dirs) {
            const readmePath = path.join(tasksDir, dir, 'README.md');
            if (fs.existsSync(readmePath)) {
                taskFiles.push(dir);
            }
            const subTasks = fs.readdirSync(path.join(tasksDir, dir))
                .filter(f => f.endsWith('.md') && !['TEMPLATE.md', 'README.md'].includes(f))
                .sort()
                .map(f => path.join(dir, f));
            taskFiles.push(...subTasks);
        }
    }

    // Build task status map
    const taskStatus = {};
    for (const taskFile of taskFiles) {
        const fullPath = path.join(tasksDir, taskFile);
        const fpath = getTaskFilePath(fullPath);

        if (!fs.existsSync(fpath)) {
            continue;
        }

        try {
            const content = fs.readFileSync(fpath, 'utf8');
            const lines = content.split('\n').slice(0, 10);

            const taskName = isDirectoryTask(fullPath) ? `${taskFile}/` : taskFile;
            let status = null;

            for (const line of lines) {
                if (line.startsWith('status:')) {
                    status = line.split(':')[1].trim();
                    break;
                }
            }

            if (status) {
                taskStatus[taskName] = status;
            }
        } catch (error) {
            continue;
        }
    }

    // Parse all index files
    const indexedTasks = {};
    const indexInfo = {};

    if (fs.existsSync(indexesDir)) {
        const indexFiles = fs.readdirSync(indexesDir)
            .filter(f => f.endsWith('.md'))
            .sort();

        for (const indexFile of indexFiles) {
            const result = parseIndexFile(path.join(indexesDir, indexFile));
            if (result) {
                const [metadata, taskLines] = result;
                if (metadata && metadata.index) {
                    const indexId = metadata.index;
                    indexInfo[indexId] = {
                        name: metadata.name || indexId,
                        description: metadata.description || '',
                        tasks: []
                    };

                    // Extract task names from the lines
                    for (const line of taskLines) {
                        // Extract task name from format: - `task-name.md` - description
                        if (line.includes('`')) {
                            try {
                                const start = line.indexOf('`') + 1;
                                const end = line.indexOf('`', start);
                                const task = line.substring(start, end);
                                if (taskStatus[task]) {
                                    indexedTasks[task] = indexId;
                                    indexInfo[indexId].tasks.push(task);
                                }
                            } catch (error) {
                                // Malformed line, skip it
                                continue;
                            }
                        }
                    }
                }
            }
        }
    }

    // Build output
    let output = "No active task set. Available tasks:\n\n";

    // Display tasks grouped by index
    const sortedIndexes = Object.entries(indexInfo).sort((a, b) => a[0].localeCompare(b[0]));
    for (const [indexId, info] of sortedIndexes) {
        if (info.tasks.length > 0) {  // Only show indexes with tasks
            output += `## ${info.name}\n`;
            if (info.description) {
                output += `${info.description}\n`;
            }
            for (const task of info.tasks) {
                if (taskStatus[task]) {
                    output += `  • ${task} (${taskStatus[task]})\n`;
                }
            }
            output += "\n";
        }
    }

    // Show unindexed tasks
    const unindexed = Object.keys(taskStatus).filter(task => !indexedTasks[task]);
    if (unindexed.length > 0) {
        output += "## Uncategorized\n";
        for (const task of unindexed.sort()) {
            output += `  • ${task} (${taskStatus[task]})\n`;
        }
        output += "\n";
    }

    // Add startup instructions
    output += `To select a task:
- Type in one of your startup commands: ${JSON.stringify(CONFIG.trigger_phrases.task_startup)}
- Include the task file you would like to start using \`@\`
- Hit Enter to activate task startup
`;

    return output;
}

//-//

// ===== EXECUTION ===== //

async function main() {
    // Read stdin for hook input
    const hookInput = await readStdin();

    //!> 1. Clear flags and todos for new session
    await editState((s) => {
        s.flags.clearFlags();
        s.todos.clearActive();
        const restored = s.todos.restoreStashed();
        STATE = s;

        context += "Cleared session flags and active todos for new session.\n\n";

        if (restored > 0) {
            context += `Restored ${restored} stashed todos from previous session:\n\n${JSON.stringify(s.todos.active)}\n\nTo clear, use \`cd .claude/hooks && node -e "const {editState} = require('./shared_state.js'); editState(s => s.todos.clearStashed()).then(()=>process.exit(0))"\`\n\n`;
        }
    });
    //!<

    //!> 2. Nuke transcripts dir
    const transcriptsDir = path.join(sessionsDir, 'transcripts');
    if (fs.existsSync(transcriptsDir)) {
        try {
            fs.rmSync(transcriptsDir, { recursive: true, force: true });
        } catch (error) {
            // Ignore errors
        }
    }
    //!<

    //!> 3. Load current task or list available tasks
    // Check for active task
    const taskFile = STATE.current_task?.filePath;
    if (taskFile && fs.existsSync(taskFile)) {
        // Check if task status is pending and update to in-progress
        let taskContent = fs.readFileSync(taskFile, 'utf8');
        let taskUpdated = false;

        // Parse task frontmatter to check status
        if (taskContent.startsWith('---')) {
            const lines = taskContent.split('\n');
            for (let i = 1; i < lines.length; i++) {
                if (lines[i] === '---') {
                    break;
                }
                if (lines[i].startsWith('status: pending')) {
                    lines[i] = 'status: in-progress';
                    taskUpdated = true;
                    // Write back the updated content
                    fs.writeFileSync(taskFile, lines.join('\n'));
                    taskContent = lines.join('\n');
                    break;
                }
            }

            // Output the full task state
            context += `Current task state:
\`\`\`json
${JSON.stringify(STATE.current_task.task_state, null, 2)}
\`\`\`

Loading task file: ${STATE.current_task.file}
${"=".repeat(60)}
${taskContent}
${"=".repeat(60)}

`;
        }

        context += `Since you are resuming an in-progress task, follow these instructions:

    1. Analyze the task requirements and work completed thoroughly
    2. Analyze any next steps itemized in the task file and, if necessary, ask any questions from the user for clarification.
    3. Propose implementation plan with structured format:

\`\`\`markdown
[PLAN: Implementation Approach]
Based on the task requirements, I propose the following implementation:

□ [Specific action 1]
  → [Expanded explanation of what this involves]

□ [Specific action 2]
  → [Expanded explanation of what this involves]

□ [Specific action 3]
  → [Expanded explanation of what this involves]

To approve these todos, you may use any of your implementation mode trigger phrases:
${JSON.stringify(CONFIG.trigger_phrases.implementation_mode)}
\`\`\`

3. Iterate based on user feedback until approved
4. Upon approval, convert proposed todos to TodoWrite exactly as written

**IMPORTANT**: Until your todos are approved, you are seeking the user's approval of an explicitly proposed and properly explained list of execution todos. Besides answering user questions during discussion, your messages should end with an expanded explanation of each todo, the clean list of todos, and **no further messages**.

**For the duration of the task**:
- Discuss before implementing
- Constantly seek user input and approval

Once approved, remember:
- *Immediately* load your proposed todo items *exactly* as you proposed them using ToDoWrite
- Work logs are maintained by the logging agent (not manually)

After completion of the last task in any todo list:
- *Do not* try to run any write-based tools (you will be automatically put into discussion mode)
- Repeat todo proposal and approval workflow for any additional write/edit-based work`;
    } else {
        context += listOpenTasksGrouped();
    }
    //!<

    //!> 4. Check cc-sessions version
    try {
        // Check local version from package.json
        let currentVersion = null;
        const pkgPath = path.join(__dirname, '..', '..', '..', 'package.json');
        if (fs.existsSync(pkgPath)) {
            const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
            currentVersion = pkg.version;
        }

        // Check npm registry for latest version
        await new Promise((resolve) => {
            https.get('https://registry.npmjs.org/cc-sessions/latest', {
                timeout: 2000
            }, (res) => {
                let data = '';
                res.on('data', chunk => data += chunk);
                res.on('end', () => {
                    try {
                        const latest = JSON.parse(data);
                        const latestVersion = latest.version;
                        if (currentVersion && currentVersion !== latestVersion) {
                            context += `Update available for cc-sessions: ${currentVersion} → ${latestVersion}. Run \`npm install -g cc-sessions\` to update.`;
                        }
                        resolve();
                    } catch (e) {
                        resolve();
                    }
                });
            }).on('error', () => resolve());

            // Timeout after 2 seconds
            setTimeout(resolve, 2000);
        });
    } catch (error) {
        // Ignore version check errors
    }
    //!<

    //-//

    // Output in JSON format as expected by Claude Code hooks
    const output = {
        hookSpecificOutput: {
            hookEventName: "SessionStart",
            additionalContext: context
        }
    };

    console.log(JSON.stringify(output));
    process.exit(0);
}

// Run main function
main().catch(error => {
    console.error(JSON.stringify({
        hookSpecificOutput: {
            hookEventName: "SessionStart",
            additionalContext: `Error in session start hook: ${error.message}`
        }
    }));
    process.exit(1);
});
